add loongarch support for upstream

yetist opened the pull request:
拟向上游提交的第一版补丁，增加 LoongArch 支持，可编译启动到 grub 界面。

注：loader 实质功能先不提交（等物理地址和efistub内核完成)，如要测试请参考 `dev-la64` 分支。

编译见最后一条提交，这里 review 通过后，这次提交会删掉再发往上游。

xen0n commented:
按照 DCO 的机制设计，Signed-off-by 必须用真名，zeiwai 是真名吗？

yetist commented:
> 按照 DCO 的机制设计，Signed-off-by 必须用真名，zeiwai 是真名吗？

已处理

xen0n commented on include/grub/loongarch64/linux.h:
```diff
@@ -0,0 +1,40 @@
+#ifndef GRUB_LOONGARCH64_LINUX_HEADER
+#define GRUB_LOONGARCH64_LINUX_HEADER 1
+
+#define GRUB_LINUX_LOONGARCH_MAGIC_SIGNATURE 0x4C413634 /* 'LA64' */
```
小端表示 4 个字节的字符串的话，应该反过来吧？`0x3436414C` 在内存里才是 `'LA64'`

yetist replied:
这个其实没用上。

xen0n replied:
没弄完建议去掉，一上来不是就说了先不做 Linux 嘛

xen0n commented on include/grub/loongarch64/linux.h:
```diff
@@ -0,0 +1,40 @@
+
+#define GRUB_LINUX_LOONGARCH_MAGIC_SIGNATURE 0x4C413634 /* 'LA64' */
+
+struct linux_loongarch64_kernel_header
+{
+  grub_uint32_t code0;		/* Executable code */
+  grub_uint32_t code1;		/* Executable code */
+  grub_uint64_t text_offset;	/* Image load offset */
+  grub_uint64_t res0;		/* reserved */
+  grub_uint64_t res1;		/* reserved */
+  grub_uint64_t res2;		/* reserved */
+  grub_uint64_t res3;		/* reserved */
+  grub_uint64_t res4;		/* reserved */
+  grub_uint32_t magic;		/* Magic number, little endian, "LA64" */
+  grub_uint32_t hdr_offset;	/* Offset of PE/COFF header */
+};
```
这个东西对应的 Linux 规范文档在哪？

yetist replied:
同上，efistub 内核的还没做完。

xen0n commented on include/grub/loongarch64/time.h:
```diff
@@ -0,0 +1,28 @@
+static inline void
+grub_cpu_idle(void)
+{
+  __asm__ __volatile__("idle 0\n");
```
这个 `\n` 不需要吧？

xen0n commented on include/grub/loongarch64/efi/memory.h:
```diff
@@ -0,0 +1,32 @@
+grub_efi_max_usable_address(void)
+{
+  grub_uint64_t addr;
+  asm volatile ("csrrd %0, 0x181" : "=r" (addr));
```
magic number 不考虑 #define 出来然后这边用立即数 operand constraint 传进来吗？

yetist replied:
这个 0x181 我也不知道是啥，文档貌似没有找到

xen0n replied:
提交上游前还是问一下。。否则我这里都过不去 ;-)

zwaizwai replied:
0x180-0x184直接含义就是cpu窗口映射地址的配置.
由uefi设置,其中0x180里面的内容是uncached地址窗口配置,内容是0x8000000000000001.0x181的内容是cached地址窗口配置0x9000000000000011.
这里也可以采用直接类似宏定义的做法#define CACHED_MEMORY_ADDR 0x9000000000000000 ,然后 grub_efi_max_usable_address 直接{ return CACHED_MEMORY_ADDR | 0xffffffffff ; }
这样做只是单纯的保证后面uefi配置时改成了别的前缀后(比如0x12xx,0x31xx)还可以兼容运行,没什么特别的意义.

xen0n replied:
> 0x180-0x184直接含义就是cpu窗口映射地址的配置. 由uefi设置,其中0x180里面的内容是uncached地址窗口配置,内容是0x8000000000000001.0x181的内容是cached地址窗口配置0x9000000000000011. 这里也可以采用直接类似宏定义的做法#define CACHED_MEMORY_ADDR 0x9000000000000000 ,然后 grub_efi_max_usable_address 直接{ return CACHED_MEMORY_ADDR | 0xffffffffff ; } 这样做只是单纯的保证后面uefi配置时改成了别的前缀后(比如0x12xx,0x31xx)还可以兼容运行,没什么特别的意义.

可以，但这些内容是哪个文档写的？应该有注释说明。

yetist replied:
@zwaizwai  那这里是否可以和其它架构一样，直接返回个地址，比如：
```
#define GRUB_EFI_MAX_USABLE_ADDRESS 0x900000ffffffffffULL
```

xen0n replied:
无论写死还是动态读（我个人显然倾向动态读），都必须附上注释

yetist replied:
@zwaizwai @xen0n

```
$ grep Address /proc/cpuinfo 
Address sizes           : 48 bits physical, 48 bits virtual
```

40位不对，3a5000 现在已经是 48 位了，所以应该和 arm64/riscv64 一样，直接设置为 `#define GRUB_EFI_MAX_USABLE_ADDRESS 0xffff ffff ffffULL` 吧？

yetist replied:
固件和内核推上游的代码，都会是物理地址，所以 grub 这里也可以直接写 0xffff ffff ffff。
但是存在一个问题，如果有主板上的固件没有刷支持物理地址的固件，grub 会起不来；而目前从 0x181 读出来的地址，在这里会返回  0x9000 00ff ffff ffff，如果有人把这个地址当成物理地址使用，就会出问题。

@zwaizwai @xen0n @kilaterlee 得看看这里怎么处理比较合适

kilaterlee replied:
我确认一下181里面是不是可以清，如果可以清，那物理地址BIOS这里清掉，你需要做兼容

xen0n replied:
按我的理解，新世界肯定是物理地址，这里如果有必要做任何兼容，那都是为了在未升级的旧世界固件可以正常运行？那么有可能通过某种方式传达这个“传参姿势版本”的信息吗？这样尽量减少猜的成分，有助于避免后人踩坑。

yetist replied:
其实也可以不做兼容，新世界使用grub-2.11+，旧世界使用grub-2.06-即可。

xen0n commented on include/grub/loongarch64/efi/memory.h:
```diff
@@ -0,0 +1,32 @@
+{
+  grub_uint64_t addr;
+  asm volatile ("csrrd %0, 0x181" : "=r" (addr));
+  return addr |= 0xffffffffffUL;
```
首先不需要赋值回去，其次这里是 mask 到 40-bit，解释一下为何不是别的值？
```suggestion
  return addr | 0xffffffffffUL;
```

zwaizwai replied:
内核的默认配置都是40bit,所以这里直接写了,很有问题的.
可以用个宏替代.

xen0n commented on include/grub/efi/api.h:
```diff
@@ -1731,7 +1731,7 @@ typedef struct grub_efi_rng_protocol grub_efi_rng_protocol_t;
 #if (GRUB_TARGET_SIZEOF_VOID_P == 4) || defined (__ia64__) \
   || defined (__aarch64__) || defined (__MINGW64__) || defined (__CYGWIN__) \
-  || defined(__riscv)
+  || defined(__riscv) || defined (__loongarch64)
```
其他跟 EFI 相关的地方，ifdef guard 都采用的 `__loongarch__` 这里为啥是 `__loongarch64`？我看好像跟位数没关系

xen0n commented on grub-core/kern/loongarch64/efi/init.c:
```diff
@@ -0,0 +1,77 @@
+#include <grub/efi/efi.h>
+#include <grub/loader.h>
+
+#define EFI_TIMER_PERIOD_MILLISECONDS(Milliseconds) ((grub_uint64_t)(Milliseconds * 10000))
```
这代码从 EDK2 抄的？参数名字又长，又以大写字母开头，感觉 `ms` 就够了

yetist replied:
> 另外第一波不实现 Linux loading 的话，是不是 Linux 相关的头文件也可以先去掉？

不能去，这个文件会 include 它：include/grub/efi/efi.h

xen0n commented on grub-core/kern/loongarch64/dl_helper.c:
```diff
@@ -0,0 +1,217 @@
+  grub_loongarch64_stack_t stack;
+
+  stack = (grub_loongarch64_stack_t) grub_malloc (sizeof(struct grub_loongarch64_stack));
+  stack->data = (grub_uint64_t*) grub_malloc (n * sizeof (grub_uint64_t));
```
有必要动态分配吗？我看现在两处下游使用都是传的 size=16，还有一处 destroy 写法有问题，建议没有特殊需求的话改回静态分配

xen0n commented on util/grub-mkimagexx.c:
```diff
@@ -784,6 +785,8 @@ SUFFIX (relocate_addrs) (Elf_Ehdr *e, struct section_metadata *smd,
   grub_uint64_t *gpptr = (void *) (pe_target + got_off);
   unsigned unmatched_adr_got_page = 0;
+  grub_loongarch64_stack_t stack;
+  stack = grub_loongarch64_stack_new (16);
```
这样写，所有架构进来都要初始化一遍这个变量，感觉不太优雅，如果静态分配空间的话倒还好（至少不涉及 malloc 只是浪费一点栈空间）

xen0n commented on util/grub-mkimagexx.c:
```diff
@@ -1422,6 +1448,9 @@ SUFFIX (relocate_addrs) (Elf_Ehdr *e, struct section_metadata *smd,
 	  }
       }
+#ifdef MKIMAGE_ELF64
+  grub_loongarch64_stack_destroy (stack);
```
这里写法就错了。。前面 new 的地方没有 ifdef guard，这里有，这样如果不是 `MKIMAGE_ELF64` 就不会执行到了

xen0n commented on util/grub-mkimagexx.c:
```diff
@@ -1668,6 +1697,52 @@ translate_relocation_pe (struct translate_context *ctx,
+	case R_LARCH_SOP_POP_32_S_5_20:
+	case R_LARCH_SOP_POP_32_S_0_5_10_16_S2:
+	case R_LARCH_SOP_POP_32_S_0_10_10_16_S2:
+		break;
```
写个注释说明不做处理的原因？

xen0n commented on grub-core/lib/loongarch64/setjmp.S:
```diff
@@ -0,0 +1,68 @@
+	st.d $ra, $a0, 0x58
+
+	move $a0, $zero
+	jirl $zero, $ra, 0
```
别的地方都 `jr $ra` 了，这里还写完整，建议都改掉（上游 binutils 从一开始就支持 `jr` 伪指令）

xen0n commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -0,0 +1,104 @@
+#include <grub/i18n.h>
+#include <grub/cpu/reloc.h>
+
+#define RELOC_STACK_MAX 1024
```
这个目前没用到也没去掉，我猜应该是挪到定义栈操作的那个 h 文件然后改成 16 这样

xen0n commented on grub-core/kern/loongarch64/cache_flush.S:
```diff
@@ -0,0 +1,44 @@
+ * a2 - line size
+ */
+FUNCTION(grub_arch_invalidate_icache_range)
+	ibar 0
```
两个函数都写下为啥不用管传入的范围参数？是硬件维护一致性所以不需要，还是 todo？

yetist replied:
咱们硬件貌似没实现进行这么精细操作的指令。

xen0n replied:
按照手册描述，`dbar 0`、`ibar 0` 之后的指令一定能看到栅障之前所有指令的访存效果，所以我们这边应该是属于“硬件维护一致性因此不需要做更多”的情况，建议写一行注释。

xen0n commented on grub-core/kern/loongarch64/cache.c:
```diff
@@ -0,0 +1,63 @@
+grub_arch_sync_dma_caches (volatile void *address __attribute__((unused)),
+			   grub_size_t len __attribute__((unused)))
+{
+  /* DMA incoherent devices not supported yet */
```
一般好像是 non-coherent DMA 这样叫

```suggestion
  /* DMA non-coherent devices not supported yet */
```

xen0n commented on grub-core/kern/loongarch64/cache.c:
```diff
@@ -0,0 +1,63 @@
+{
+  /* TODO */
+  dlinesz = 32;
+  ilinesz = 32;
```
提交前 TODO 要改掉吗？

yetist replied:
这个文件整个是抄的 riscv

xen0n replied:
可还行，那改 64 吧，至少 3A5000 的 cache line 是 64 字节的。

yetist replied:
都没啥用，调到最后都是ibar/dbar 0

xen0n replied:
那既然没用是不是可以考虑第一版都去了。。。

xen0n commented:
另外第一波不实现 Linux loading 的话，是不是 Linux 相关的头文件也可以先去掉？

yetist commented:
谁帮忙看看类似警告为啥在 x86 上没有，在 la 上就有：
```
../grub-core/lib/libgcrypt-grub/cipher/md5.c:320:20: 警告：类型转换增大了目标类型的对齐需求 [-Wcast-align]  
```

yetist commented:
更新了一把，看看还有啥问题需要处理吗？

xen0n commented on grub-core/kern/efi/mm.c:
```diff
@@ -659,7 +659,8 @@ grub_efi_mm_init (void)
 
-#if defined (__aarch64__) || defined (__arm__) || defined (__riscv)
+#if defined (__aarch64__) || defined (__arm__) || defined (__riscv) \
+|| defined (__loongarch__)
```
`||` 放上一行结尾？

xen0n commented on grub-core/kern/loongarch64/cache_flush.S:
```diff
@@ -0,0 +1,32 @@
+	.file	"cache_flush.S"
+	.text
+/*
+ * Hardware maintenance cache consistency does not need to do more
```
Broken English.

```suggestion
 * No further work to do because cache consistency is maintained by hardware on
 * LoongArch.
```

xen0n commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -0,0 +1,100 @@
+	GRUB_LOONGARCH64_RELOCATION (&stack, place, sym_addr)
+	default:
+	  {
+	    char rel_info[17]; /* log16(2^64) = 16, plus NUL. */
```
```suggestion
	    char rel_info[17]; /* log16(2^64) = 16, plus NUL.  */
```

xen0n commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -0,0 +1,100 @@
+#pragma GCC diagnostic ignored "-Wcast-align"
+
+/*
+ * Unified function for both REL and RELA
```
```suggestion
 * Unified function for both REL and RELA.
```

xen0n commented on grub-core/kern/loongarch64/dl.c:
```diff
@@ -0,0 +1,100 @@
+
+      if (rel->r_offset >= seg->size)
+	return grub_error (GRUB_ERR_BAD_MODULE,
+			   "reloc offset is out of the segment");
```
```suggestion
			   "reloc offset is outside the segment");
```

xen0n commented on grub-core/lib/loongarch64/setjmp.S:
```diff
@@ -0,0 +1,68 @@
+
+/*
+   int grub_setjmp (jmp_buf env)
+*/
```
Use proper comment syntax:

```
/*
 * int grub_setjmp (jmp_buf env)
 */
```

xen0n commented on grub-core/lib/loongarch64/setjmp.S:
```diff
@@ -0,0 +1,68 @@
+
+/*
+   void grub_longjmp (jmp_buf env, int val)
+*/
```
Similarly here.

xen0n commented on include/grub/loongarch64/efi/memory.h:
```diff
@@ -0,0 +1,24 @@
+#ifndef GRUB_MEMORY_CPU_HEADER
+#include <grub/efi/memory.h>
+
+#define GRUB_EFI_MAX_USABLE_ADDRESS 0xffffffffffffULL
```
还是加一行注释说明 48 位的来历

yetist replied:
include/grub/*/efi/memory.h 大家都没写，可以不写。

xen0n replied:
如果都没写那好像也行

xen0n commented on include/grub/loongarch64/reloc.h:
```diff
@@ -0,0 +1,114 @@
+
+typedef struct grub_loongarch64_stack* grub_loongarch64_stack_t;
+
+#define grub_loongarch64_stack_init(name, n)	\
+  do {						\
+      name.top = -1;				\
+      name.count = n;				\
+        if (n > LOONGARCH64_STACK_MAX)		\
+	  name.count = LOONGARCH64_STACK_MAX;	\
+  } while(0);
```
弄成一个宏相比一个函数有什么好处？感觉好像没有太大必要。。

yetist replied:
1. 原来是动态分配，使用 new(n)，可以在内部封装top和count，并判断 n 是否超范围，这个没有问题。

2. 现在使用宏，可以这样使用：
```
  struct grub_loongarch64_stack stack;
  grub_loongarch64_stack_init (stack, 12);
```
代码也干净，后续的宏都是针对 stack 这个变量进行处理。

而如果使用函数，比如 `grub_loongarch64_stack_init (n)`，那如何得知在调用的地方声明的`struct grub_loongarch64_stack` 类型变量是叫 a 还是叫 b 呢？除非在函数里直接写死。
如果写死它叫 a，并且在  `grub_loongarch64_stack_XXX ()` 系列函数中就处理a，那么问题来了，如果有多处调用和处理的话，栈直接乱掉。

xen0n replied:
不是这样啊，函数签名写成 `void grub_loongarch64_stack_init (grub_loongarch64_stack_t s);` 就行了，里面操作 `s->top` `s->count` 即可

然后我是觉得目前自定义 stack size 的能力也可以先去掉，写死 `LOONGARCH64_STACK_MAX` 问题也不大的。

xen0n commented on include/grub/loongarch64/types.h:
```diff
@@ -0,0 +1,34 @@
+/* LoongArch is little-endian.  */
+#undef GRUB_TARGET_WORDS_BIGENDIAN
+
+/* Unaligned accesses only supported if MMU enabled */
```
```suggestion
/* Unaligned accesses are only supported if MMU is enabled.  */
```

cnmushiba commented on grub-core/kern/loongarch64/efi/startup.S:
```diff
@@ -0,0 +1,34 @@
+	la		$a2, EXT_C(grub_efi_system_table)
+	st.d		$a1, $a2, 0
+
+	bl		EXT_C(grub_main)
```
grub_main好像不返回，是不是直接`bl` -> `b`就可以了？

yetist replied:
好像有道理啊，下一版本修改过来吧，现在这个已经给上游发邮件了：https://lists.gnu.org/archive/html/grub-devel/2022-01/msg00036.html

xen0n commented on include/grub/loongarch64/reloc.h:
```diff
@@ -0,0 +1,107 @@
+
+#define LOONGARCH64_STACK_MAX 16
+
+struct grub_loongarch64_stack
+{
+  grub_uint64_t data[LOONGARCH64_STACK_MAX];
+  int count;
+  int top;
+};
```
这个结构体的所有字段现在都相当于“私有”了，考虑收进实现的 c 文件，这里只留一个 `struct grub_loongarch64_stack;` 如何？

yetist replied:
收进c的话，在声明这种变量的地方无法就知道结构体大小了，会报错。

xen0n replied:
哦哦，对，确实，那这里不用动了

xen0n commented on include/grub/loongarch64/linux.h:
```diff
@@ -0,0 +1,38 @@
+#ifndef GRUB_LOONGARCH64_LINUX_HEADER
+#define GRUB_LOONGARCH64_LINUX_HEADER 1
+
+struct linux_loongarch64_kernel_header
+{
+  grub_uint32_t code0;		/* Executable code */
+  grub_uint32_t code1;		/* Executable code */
+  grub_uint64_t text_offset;	/* Image load offset */
+  grub_uint64_t res0;		/* reserved */
+  grub_uint64_t res1;		/* reserved */
+  grub_uint64_t res2;		/* reserved */
+  grub_uint64_t magic;		/* Magic number */
+  grub_uint64_t res3;		/* reserved */
+  grub_uint32_t res4;		/* reserved */
+  grub_uint32_t hdr_offset;	/* Offset of PE/COFF header */
+};
+
+#define linux_arch_kernel_header linux_loongarch64_kernel_header
```
这部分是编译 `linux` 命令 stub 也需要的吗？否则建议这一版也干掉，感觉上就不是什么规范化的东西

yetist replied:
结构体得定义出来，否则编译会出错的。
```
include/grub/efi/efi.h:grub_err_t grub_arch_efi_linux_check_image(struct linux_arch_kernel_header *lh);
```

xen0n replied:
那先把实质内容干掉？我理解这些在你第一阶段都是不需要的，或者从常识上看，也应该随着 Linux EFI stub 结构的确定而随着第二波一起提交。

yetist replied:
改了，reviewed-by 把你加上，你邮箱要用github这个吗？

xen0n replied:
`Reviewed-by: WANG Xuerui <git@xen0n.name>`

xen0n reviewed (commented):
除了这回的两个小坑之外，我觉得没啥问题了

xen0n reviewed (approved):
想起来 Reviewed-by 在你这应该先不加，等你发邮件列表了 cc 带上我，我去回 For whole series: Reviewed-by 会比较好。

yetist commented:
@zwaizwai @kilaterlee 

这个补丁你们还有啥意见没？

xen0n commented:
`LoongArch: Add Linux load logic` 这个 commit message 不实（实际只是加了 `linux` `initrd` 命令的 stub 实现），建议改一下 `LoongArch: Add stubs for Linux loading commands`

yetist committed:
PE: Add LoongArch definitions

Added LoongArch machine type[1] and base relocation type[2] to PE file format

Machine Types:
- IMAGE_FILE_MACHINE_LOONGARCH32 0x6232
- IMAGE_FILE_MACHINE_LOONGARCH64 0x6264

Base Relocation Types:
- IMAGE_REL_BASED_LOONGARCH32_MARK_LA 8
- IMAGE_REL_BASED_LOONGARCH64_MARK_LA 8

[1]: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#machine-types
[2]: https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#base-relocation-types

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
Add LoongArch definitions

Add e_machine[1] and relocations types[2] in ELF for LoongArch.

[1]: https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_e_machine_identifies_the_machine
[2]: https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_relocations

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
LoongArch: Add setjmp implementation

This patch adds a setjmp implementation for LoongArch.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Sun Haiyong <sunhaiyong@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
LoongArch: Add early startup code

On entry, we need to save the system table pointer as well as our image
handle. Add an early startup file that saves them and then brings us
into our main function.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
LoongArch: Add awareness for LoongArch relocations

This patch adds awareness of LoongArch relocations throughout the grub tools
as well as dynamic linkage and elf->PE relocation conversion support.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
LoongArch: Add auxiliary files

Add some auxiliary functions to the LoongArch architecture to support
memory, cache, timers, etc.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist committed:
LoongArch: Add to build system

This patch adds support for LoongArch to the grub build system. With this
patch, I can successfully build grub on LoongArch as a UEFI application.

Signed-off-by: Zhou Yang <zhouyang@loongson.cn>
Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist commented:
更新到v3

yetist committed:
tests: add support for LoongArch

The test results of running "make check" with qemu 7.2 are as follows:
============================================================================
Testsuite summary for GRUB 2.11
============================================================================
============================================================================

These ERROR cases need to be run as root user:
  ERROR: ext234_test
  ERROR: hfsplus_test
  ERROR: ntfs_test
  ERROR: reiserfs_test
  ERROR: fat_test
  ERROR: minixfs_test
  ERROR: xfs_test
  ERROR: f2fs_test
  ERROR: nilfs2_test
  ERROR: romfs_test
  ERROR: exfat_test
  ERROR: udf_test
  ERROR: hfs_test
  ERROR: jfs_test
  ERROR: btrfs_test
  ERROR: zfs_test
  ERROR: luks1_test
  ERROR: luks2_test
  ERROR: grub_cmd_cryptomount

These test cases are skipped:
  SKIP: pata_test
  SKIP: ahci_test
  SKIP: uhci_test
  SKIP: ohci_test
  SKIP: ehci_test
  SKIP: fddboot_test
  SKIP: netboot_test
  SKIP: pseries_test

This test case fails because of: qemu-system-loongarch64: ram_size must be greater than 1G.
  FAIL: grub_func_test

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist commented:
更新到v5。
修复make distcheck
自动检测CFLAGS参数
增加 tests 支持，测试结果见附件。
[test-suite.log](https://github.com/loongarch64/grub/files/9069024/test-suite.log)

xen0n commented on tests/util/grub-shell.in:
```diff
@@ -194,6 +194,15 @@ case "${grub_modinfo_target_cpu}-${grub_modinfo_platform}" in
+	boot=hd
+	console=console
+	trim=1
+	qemuopts="-machine virt -cpu la464 -smp 2 -nographic -bios /usr/share/qemu/edk2-loongarch64-code.fd $qemuopts"
```
`-bios xxx` 不该进上游吧？一方面现在还没有 LA 固件二进制进 qemu 上游，另一方面可能到时候也不叫这个名。

yetist replied:
这个是预期进上上游的文件名及路径。

另外，你看一下其它架构的情况，如果这部分你来写的话，你准备写成什么样？

yetist replied:
希望能给点更好的建议，谢谢

yetist replied:
我从qemu源码编译安装出来，bios文件的情况是这样子：
```
$ ls -ltr /usr/share/qemu/*fd
-rw-r--r-- 1 root root  3653632  7月  5 15:10 /usr/share/qemu/edk2-i386-code.fd
-rw-r--r-- 1 root root 67108864  7月  5 15:10 /usr/share/qemu/edk2-arm-code.fd
-rw-r--r-- 1 root root 67108864  7月  5 15:10 /usr/share/qemu/edk2-arm-vars.fd
-rw-r--r-- 1 root root 67108864  7月  5 15:10 /usr/share/qemu/edk2-aarch64-code.fd
-rw-r--r-- 1 root root   540672  7月  5 15:10 /usr/share/qemu/edk2-i386-vars.fd
-rw-r--r-- 1 root root  3653632  7月  5 15:10 /usr/share/qemu/edk2-i386-secure-code.fd
-rw-r--r-- 1 root root  3653632  7月  5 15:10 /usr/share/qemu/edk2-x86_64-secure-code.fd
-rw-r--r-- 1 root root  3653632  7月  5 15:10 /usr/share/qemu/edk2-x86_64-code.fd
```
所以，预期上游默认的文件应该为 `/usr/share/qemu/edk2-loongarch64-code.fd`。

而不同发行版可以自行决定这个文件安装到哪里，叫什么名字，这种情况只能选择忽略掉。

yetist replied:
@loongarch64/dev-team 

如果跑测试用例，进行验证的话，请使用：

qemu代码：https://github.com/loongson/qemu， 分支：tcg-full-system

bios：https://github.com/yangxiaojuan-loongson/qemu-binary/blob/main/binary-0711/QEMU_EFI.fd, 需要复制为 `/usr/share/qemu/edk2-loongarch64-code.fd`

yetist replied:
@xen0n 

白老师，有条件帮忙验证吗？

xen0n replied:
> @xen0n
> 
> 白老师，有条件帮忙验证吗？

晚点

yetist committed:
LoongArch: Support new relocation types in v2.00 ABI

After LoongArch upgraded[1] the EFI ABI document from 1.0 to 2.0[2], a new set of
relocation types (starting from number 64) was added to replace the 1.0 abi
relocation. It is implemented in binutils 2.40 and gcc 13.

This patch adds support for abi 2.0 relocation, and plans to remove the code
related to 1.0 relocation in future versions.

[1]: https://github.com/loongson/LoongArch-Documentation/pull/57
[2]: https://loongson.github.io/LoongArch-Documentation/LoongArch-ELF-ABI-EN.html#_appendix_revision_history

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

yetist commented:
Fix make distcheck failed and rebase on the last commit.

Tested with qemu-7.1.0-rc0, the test results:
```
=================================
   GRUB 2.11: ./test-suite.log
=================================

# TOTAL: 82
# PASS:  52
# SKIP:  8
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 20
```                                                                                 
[test-suite.log](https://github.com/loongarch64/grub/files/9216733/test-suite.log)

yetist commented:
Support new relocation types.

Tested with qemu-7.1.0-rc0, the test results:
```
============================================================================
Testsuite summary for GRUB 2.11
============================================================================
# TOTAL: 82
# PASS:  56
# SKIP:  8
# XFAIL: 0
# FAIL:  2
# XPASS: 0
# ERROR: 16
============================================================================
See ./test-suite.log
Please report to bug-grub@gnu.org
============================================================================
```
[test-suite.log](https://github.com/loongarch64/grub/files/9239702/test-suite.log)

There are two test case failed:
FAIL: squashfs_test
FAIL: grub_func_test

The squashfs_test failed because FSTIME does not match:

```
Device proc: Filesystem type procfs - Sector size 512B - Total size 0KiB         
Device loop0: Filesystem type squash4 - Last modification time 2022-08-02 06:21:25 Tuesday - Sector size 512B - Total size 10680KiB
Device host: Filesystem type hostfs - Sector size 512B - Total size 0KiB         
                                                                                 
FSTIME FAIL                                                                      
2022-08-02 14:21:25                                                              
Device loop0: Filesystem type squash4 - Last modification time 2022-08-02 06:21:25 Tuesday - Sector size 512B - Total size 10680KiB
FAIL squashfs_test (exit status: 1)                                              
```

This failure can be ignored because my timezone is UTC+8:
```
$ timedatectl
               Local time: 二 2022-08-02 14:47:45 CST
           Universal time: 二 2022-08-02 06:47:45 UTC
                 RTC time: 二 2022-08-02 06:47:46
                Time zone: Asia/Shanghai (CST, +0800)
System clock synchronized: yes
              NTP service: active
          RTC in local TZ: no
```

yetist committed:
tests: fix squashfs_test failed to check fstime

Using the "unsquashfs -s" command to get the time from the image will cause the test case to fail if the local timezone is used.

This patch makes "unsquashfs" return UTC time to fix this bug:

  unsquashfs -s /tmp/grub-fs-tester.20230407111703613257436.squash4_gzip.9R4/squash4_gzip_512_4096_1_0.img
  grep '^Creation'
  awk '{print $6 " " $7 " " $8 " " $9 " " $10; }'
  FSTIME='Fri Apr 7 11:17:05 2023'
  date -d 'Fri Apr 7 11:17:05 2023' -u '+%Y-%m-%d %H:%M:%S'
  FSTIME='2023-04-07 11:17:05'
  date -d '2023-04-07 11:17:05 UTC -1 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM1='2023-04-07 11:17:04'
  date -d '2023-04-07 11:17:05 UTC -2 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM2='2023-04-07 11:17:03'
  date -d '2023-04-07 11:17:05 UTC -3 second' -u '+%Y-%m-%d %H:%M:%S'
  FSTIMEM3='2023-04-07 11:17:02'
  grep -F 'Last modification time 2023-04-07 11:17:05'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:04'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:03'
  echo 'Device loop0: Filesystem type squash4 - Last modification time 2023-04-07 03:17:05 Friday - Sector size 512B - Total size 10680KiB'
  grep -F 'Last modification time 2023-04-07 11:17:02'
  echo FSTIME FAIL

Signed-off-by: Xiaotian Wu <wuxiaotian@loongson.cn>

xen0n commented:
Upstream has [merged everything](https://git.savannah.gnu.org/gitweb/?p=grub.git;a=commitdiff;h=2188db1a9f562f0987dd47e0775c3d4d34cccd25), so this PR served its purpose well and can be closed.
